workflow/scripts/intersect/estimate_wind_fields.py

```diff
@@ -0,0 +1,279 @@
+
+    # forward azimuth angle and distances from track eye to next track eye
+    advection_azimuth_deg, _, eye_step_distance_m = geod_wgs84.inv(
+        track.geometry.x.iloc[:-1],
+        track.geometry.y.iloc[:-1],
+        track.geometry.x.iloc[1:],
+        track.geometry.y.iloc[1:],
+    )
+
+    # gapfill last period/distance values with penultimate value
+    period = track.index[1:] - track.index[:-1]
+    period = period.append(period[-1:])
+    eye_step_distance_m = [*eye_step_distance_m, eye_step_distance_m[-1]]
+    track["advection_azimuth_deg"] = [*advection_azimuth_deg, advection_azimuth_deg[-1]]
+
+    # calculate eye speed
+    track["eye_speed_ms"] = eye_step_distance_m / period.seconds.values
+
+    # hemisphere belongs to {-1, 1}
+    track["hemisphere"] = np.sign(track.geometry.y)
+
+    grid_shape: tuple[int, int] = (len(longitude), len(latitude))
+    adv_field: np.ndarray = np.zeros((len(track), *grid_shape), dtype=complex)
+    rot_field: np.ndarray = np.zeros((len(track), *grid_shape), dtype=complex)
+
+    for track_i, track_point in enumerate(track.itertuples()):
+
+        adv_vector: np.complex128 = advective_vector(
+            track_point.advection_azimuth_deg,
+            track_point.eye_speed_ms,
+            track_point.hemisphere,
+        )
+
+        adv_field[track_i, :] = np.full(grid_shape, adv_vector)
+
+        # maximum wind speed, less advective component
+        # this is the maximum tangential wind speed in the eye's non-rotating reference frame
+        max_wind_speed_relative_to_eye_ms: float = track_point.max_wind_speed_ms - np.abs(adv_vector)
+
+        rot_field[track_i, :] = rotational_field(
+            longitude,  # degrees
+            latitude,  # degrees
+            track_point.geometry.x,  # degrees
+            track_point.geometry.y,  # degrees
+            track_point.radius_to_max_winds_km * 1_000,  # convert to meters
+            max_wind_speed_relative_to_eye_ms,
+            track_point.min_pressure_hpa * 100,  # convert to Pascals
+            ENV_PRESSURE[basin] * 100,  # convert to Pascals
+        )
+
+    # sum components of wind field, (timesteps, y, x)
+    wind_field: np.ndarray[complex] = adv_field + rot_field
+
+    # hypotenuse of u,v wind vectors, (timesteps, y, x)
+    wind_speeds: np.ndarray[float] = np.abs(wind_field)
+
+    # find in which timestep the maximum speed values is for each raster pixel
+    timestep_indicies: np.ndarray[int] = np.argmax(wind_speeds, axis=0)
+
+    # take max along timestep axis, giving (y, x)
+    # equivalently, could use timestep_indicies.choose(wind_speeds), but this limited to 32 possible timesteps
+    max_wind_speeds: np.ndarray[float] = np.take_along_axis(
+        wind_speeds,
+        # N.B. take_along_axis indicies must be same rank as choices, so reshape from e.g. (50,50) to (1,50,50)
+        timestep_indicies.reshape(1, *timestep_indicies.shape),
+        axis=0
+    ).squeeze()  # drop the redundant first axis
```

Comment:
Could this all be `max_wind_speeds: np.ndarray[float] = np.max(wind_speeds, axis=0)`? I'm not sure why we need to find the indices then pick using them.

Reply:
Yep, I can't remember why I did that. Possibly for a plot that I didn't end up making. I've simplified it now.